### PR TITLE
feat(netbox): Add Netbox Script to create a Switch

### PR DESCRIPTION
**Pull request by sklirg**
An initial look at creating a switch "our way" in NetBox.

This PR is a work in progress and will contain comments and demos down the line 👍 

ToDo:

- [x] Feature-parity with [create-script.py](https://github.com/gathering/tgmanage/blob/main/examples/tg19/netbox_tools/create_switch.py)
- [x] Provide smart suggestions/dropdowns where feasible
- [x] Send data to Gondul ([will be done with NetBox webhooks](https://github.com/gathering/tgmanage/pull/106#issuecomment-1457129228))
- [x] Ensure cables are marked as 'connected' https://github.com/gathering/tgmanage/pull/106/files#r1127117814
- [x] Add docs (and maybe example implementation?) of how to [call this script using the NetBox API](https://demo.netbox.dev/static/docs/customization/custom-scripts/#via-the-api)
- [x] Clean up in options, remove options that can be inferred from others (e.g. "site" based on which "destination device" is picked)

ToDo in NetBox(es) where the scripts are to be deployed:
- [x] Create CustomField for VLAN ([implementation](https://github.com/gathering/tgmanage/commit/8c185ddbd59cb61835c615a5b7207a2b2fc99cd5))
- [x] Set up auto-sync of scripts from git

---

NetBox will filter to relevant objects, e.g. showing only devices on the "destination device", as well as filtering away already-connected interfaces.

image

### CODE REVIEW

**Comment by JonasKs** [comment]
This is cool, never knew it was this easy to customize NetBox!

**Comment by sklirg** [comment]
Current state:

https://user-images.githubusercontent.com/5422571/223259940-f8b3e439-de12-460b-b9f6-053ded473baa.mov

**Comment by sklirg** [comment]
For sending to Gondul I will simply reverse-engineer [the APIs](https://github.com/gathering/gondul/blob/master/web/api/write) and create webhooks from Netbox on "create" events. It doesn't look like Gondul is happy for _updates_, but I'm going to look a bit more into that.

File: tools/netbox/scripts/create-switch/create-switch.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            b = CableTermination.objects.create(
+                cable_end='B',
+                cable=cable,
+                termination_id=b_interface.id,
+                termination_type=interface_type,
+            )
+            cable = Cable.objects.get(id=cable.id)
+            # TODO: fix cables being 'connected'
+            #cable.a_terminations = [a]
+            #cable.b_terminations = [b]
```
Comment (sklirg) [inline tools/netbox/scripts/create-switch/create-switch.py:203]: #todo

**Comment by JonasKs** [comment]
When working with Django, you can use [atomic transactions](https://docs.djangoproject.com/en/4.1/topics/db/transactions/). I’ve only used them in API views, but I assume the same approach works around the `run` function. 

> Atomicity is the defining property of database transactions. atomic allows us to create a block of code within which the atomicity on the database is guaranteed. If the block of code is successfully completed, the changes are committed to the database. If there is an exception, the changes are rolled back.

Basically all you need to do is adding a decorator to your function. 

```py
from django.db import transaction

@transaction.atomic
def run()
```

Then, if anything goes wrong during execution, it won’t partially update the database on the `Model.save()` calls. 😊

**Comment by sklirg** [comment]
This is actually already implemented in the Custom Scripts -- before clicking "Run Script" you can uncheck "Commit" which will automatically rollback any changes. And if the script fails, it will also roll back any changes. :)

**Comment by JonasKs** [comment]
NetBox man, genius.

**Review by slinderud — COMMENTED** [review]
lgtm

**Pull request merged**